src/tree/hist/histogram.h

```diff
@@ -76,20 +79,51 @@ class HistogramBuilder {
                             std::vector<bst_node_t> const &nodes_to_build,
                             common::RowSetCollection const &row_set_collection,
                             common::Span<GradientPair const> gpair_h, bool force_read_by_column) {
-    // Parallel processing by nodes and data in each node
-    common::ParallelFor2d(space, this->n_threads_, [&](size_t nid_in_set, common::Range1d r) {
-      const auto tid = static_cast<unsigned>(omp_get_thread_num());
-      bst_node_t const nidx = nodes_to_build[nid_in_set];
-      auto elem = row_set_collection[nidx];
-      auto start_of_row_set = std::min(r.begin(), elem.Size());
-      auto end_of_row_set = std::min(r.end(), elem.Size());
-      auto rid_set = common::RowSetCollection::Elem(elem.begin + start_of_row_set,
-                                                    elem.begin + end_of_row_set, nidx);
-      auto hist = buffer_.GetInitializedHist(tid, nid_in_set);
-      if (rid_set.Size() != 0) {
-        common::BuildHist<any_missing>(gpair_h, rid_set, gidx, hist, force_read_by_column);
+    if (is_distributed_ && is_col_split_ && is_secure_) {
+      // Call the interface to transmit gidx information to the secure worker
+      // for encrypted histogram compuation
+      auto slots = std::vector<int>();
+      auto num_rows = row_set_collection[0].Size();
```

Comment:
Do we need to use the first node `0` every time?

Reply:
good catch! I think we only need to perform InitAggregationContext once, but in this implementation we did it several times, I think we can check if the context has been initialized and only perform this step at the beginning. (@nvidianz is my understanding correct?)

Reply:
updated to make it happening only once